### ISSUE TITLE
recursive static initialization is broken

### ISSUE THREAD

**Issue by solson** [solson/miri#120]
I investigated the only "tried to modify constant memory" in #55 and it's due to the fact that Miri can't properly initialize statics that recursively refer to each other. This would be low priority since it only affects the unstable `#![feature(static_recursion)]`, but the feature looks like it will be [stable soon](https://github.com/rust-lang/rust/issues/29719).

E.g.

```rust
struct S(&'static S);
static S1: S = S(&S2);
static S2: S = S(&S1);
```

When you start initializing S1:

1. you'll start initializing S2 because it's required (`ConstantExtractor`)
2. write the final result for S2
3. freeze S2 and all allocations it references (this freezes S1)
4. pick up where you left off initializing S1
5. write the final result for S1 (fails because it's frozen)

**Comment by oli-obk** [comment]
I gotta run. can you try what happens with the PR I just opened?

**Comment by oli-obk** [comment]
it's also fixed by #121
